### PR TITLE
Correct have_fastchunks arena adjustment

### PR DESCRIPTION
**Pull request by CptGibbon**
## Correct have_fastchunks arena adjustment ##

Patches a check that (though uncommented) appears to have been introduced to compensate for the `have_fastchunks` flag being moved from an arena's `flags` into its own field in GLIBC 2.27. The code in question is used to parse arenas when the linked GLIBC build has no debug symbols available. The only thing I've changed is the version number, it should read >= 2.27 rather that 2.26. In its current state, if you try to debug a program linked to a GLIBC 2.26 build without debug symbols, the arena layout is misinterpreted and heap commands will hang or give incorrect information.

Only tested on x86_64 hardware.

### CODE REVIEW

**Comment by theguy147** [comment]
Can you provide some test case that fails with the glibc version check for version >= 2.26?

I just created a testing setup with Fedora27 (that uses glibc version 2.26) and I have the exact opposite experience: Before applying your patch it works fine and afterwards it doesn't anymore.

My test setup was done with the current `generic/fedora27` vagrant box (https://app.vagrantup.com/generic/boxes/fedora27):

```bash
[vagrant@fedora27 ~]$ uname -a
Linux fedora27.localdomain 4.18.19-100.fc27.x86_64 #1 SMP Wed Nov 14 22:04:34 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

### This is without your patch:

image

### This is with your patch:

image

Note that `heap chunks` doesn't return anything anymore and `heap bins` makes GEF freeze without returning to the prompt.

As a test binary I just used this very simple program, compiled it without debug symbols and set a breakpoint after the call to `malloc`:
```c
// file: test.c
#include<stdlib.h>
#include<stdio.h>

int main() {
	int* ptr = malloc(0x100);
	printf("%p\n", ptr);
	return 0;
}
```

**Comment by CptGibbon** [comment]
I tested against GLIBC versions 2.25, 2.26 & 2.27 with & without debug symbols. In my tests the versions _with_ debug symbols performed correctly before the patch, but the 2.26 version _without_ debug symbols failed. After the patch all versions with or without debug symbols perform correctly.
I'll see if I can replicate the Fedora issue, in the mean time I would note that you're testing with debug symbols which is not what this patch is targeting.

**Comment by theguy147** [comment]
Oh, sorry, I misread. I thought you meant that the target executable should not have debug symbols. I will look into your scenario.
But yes, backporting, would definitely mess with the hardcoded version checks.

**Comment by CptGibbon** [comment]
I've tested against a Fedora 27 VM with its native GLIBC 2.26 and am still seeing my original results, it doesn't look like Red Hat backported any major arena changes to their 2.26 libc build.
Not to imply your tests are faulty 😅, but would you mind confirming you've used my patch and not changed the value manually? There are 3 of those version checks & they all look very similar, but only one of them needs changing.
It's just odd that GEF behaves differently with the patch on your setup because that code shouldn't fire if the libc debug symbols are present, which it appears they are on your system (normally you'd see a message along the lines of `missing separate debug info for target...`
If the patch is sound then I'm not sure what to do...

**Comment by theguy147** [comment]
IMO this PR should be merged after adding the missing comment to respect vanilla GLIBC versions instead of every possible backport (which is just impossible). We could then add ways of respecting backports to https://github.com/hugsy/gef-extras with e.g. some kind of glibc database with offsets of important structs for GEF (per distro and glibc version) or something.

File: gef.py [hunk 1]
```diff
@@ -648,7 +648,7 @@ def __init__(self, addr):
             ptr_type = "unsigned long" if current_arch.ptrsize == 8 else "unsigned int"
             self.size_t = cached_lookup_type(ptr_type)
 
-        if get_libc_version() >= (2, 26):
+        if get_libc_version() >= (2, 27):
```
[the remark this answers is not in the document]
Comment (CptGibbon) [reply gef.py:654]: I've added a comment, let me know if it's what you wanted.

[the remark this answers is not in the document]
Comment (theguy147) [reply gef.py:654]: thanks, that's exactly what I had in mind.

**Review by theguy147 — APPROVED** [review]

**Pull request merged**